### PR TITLE
Set up a search scope when searching for mbe macro references

### LINKED ISSUE #6184: macro rename failed
image
after i rename the macro, the macro definition has been renamed, but all the reference not change.
image

### PR DESCRIPTION
**Pull request by Veykril**
Closes #6184

### CODE REVIEW

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -1160,6 +1160,11 @@ impl MacroDef {
         // FIXME: wrong for `ProcMacro`
         matches!(self.id.kind, MacroDefKind::ProcMacro(..) | MacroDefKind::BuiltInDerive(..))
     }
+
+    /// Indicate it is a declarative macro
+    pub fn is_declarative(&self) -> bool {
+        matches!(self.id.kind, MacroDefKind::Declarative(..))
+    }
```
Comment (matklad) [inline crates/hir/src/lib.rs:1167]: Let's merge this three into `fn kind()`

File: crates/ide_db/src/search.rs [hunk 2]
```diff
@@ -256,7 +256,25 @@ impl Definition {
...
+        if let Definition::Macro(macro_def) = self {
+            if macro_def.is_declarative() {
+                return if macro_def.attrs(db).by_key("macro_export").exists() {
+                    rev_dep_scope()
+                } else {
+                    let source_root_id = db.file_source_root(file_id);
+                    let source_root = db.source_root(source_root_id);
+                    SearchScope::new(
+                        source_root.iter().map(|id| (id, None)).collect::<FxHashMap<_, _>>(),
+                    )
```
Comment (matklad) [inline crates/ide_db/src/search.rs:271]: It feels like this should more directly express "this craete" concept. Probably makes sense to look at how we handle `pub(crate)`.

Comment (Veykril) [reply crates/ide_db/src/search.rs]: From what I can tell `pub(crate)` is handled the same way as any `pub(in _)` with the root module being the module there. So there isn't much to adapt. I tried shuffling around some stuff but I'm honestly unsure if this came out better or not.

Comment (matklad) [reply crates/ide_db/src/search.rs]: Hm, yeah, I am still not completely happy with this. It feels wrong that SourceRoot has such a significant presence here, ideally, search scopes should be described in terms of `hir`, and lowered to files at the very end. 

So, I'd expact that we describe scope as just "crates X, Y and Z" and then, when lowering that to a set of files, we do either of:

* approximate the set of files by looking at the source roots containing the root files of the crates
* calculate the precise set of files by eagerly computing def maps and walking the tree of modulus. 

That being said, I am fine with any implementation for the time being, so use your judgement. 

I feel we well be re-engineering the search infrastructure many times in the future, as it is very unclear how text-based trigram index lazy approach should be married with a language with macros.

Comment (Veykril) [reply crates/ide_db/src/search.rs]: Managed to hide the SourceRoot usage a bit more from the api, now `SearchScope` only takes `hir`.

**Comment by lnicola** [comment]
changelog fix

**Pull request merged**